Carpentries Workbench (#1308)
carpentries/carpentries.org: _posts/2022/01/2022-01-31-live-lesson-infrastructure.md

---
layout: page
authors: ["Zhian N. Kamvar"]
teaser: "We announce the first draft of our accessible lesson infrastructure and call for beta testers"

> title: "First Draft of The Carpentries Workbench"

Reviewer: ```suggestion
title: "Introducing The Carpentries Workbench"
```

You mention it's a first draft in the teaser, which will appear as a subtitle on the page.

Author: I thought I added this in the previous commit, but github was being wah

date: 2022-01-31
time: "09:00:00"
tags: ["Lesson Infrastructure", "Community"]
---

## What is The Carpentries Workbench?

[The Carpentries Workbench][workbench] is a fresh redesign of our lesson

> template from the ground up that clearly separates the tools from the content.

Reviewer 2: ```suggestion
infrastructure from the ground up that clearly separates the tools from the content.
```

Author: Thought I included this in the earlier batch 🤦🏼

We hope it will better serve the needs of our community members and create
paths forward for improving access, portability, stability, and reliability of
our lesson infrastructure as the repoirtoir of our lesson developers continues
to grow.

Following the guidelines we set out in [our post on design

> principles][design-principles], The Carpentries Workbench is built with R and

Reviewer 2: add comment that people can use the template without knowing R (like they use the template now without knowing Ruby)

> Pandoc in three custom packages called [{sandpaper}], [{varnish}], and

Reviewer 2: add link to pandoc for people not familiar with it?

[{pegboard}] for deployment, styling, and validation, respectively. To find out

> more about the process we used for validating that the infrastructure, you can

Reviewer 2: ```suggestion
more about the process we used for validating the infrastructure, you can
```

read our [blog post about the alpha testing][alpha-test] and our [blog post
about the UX testing and design][ux-test], or if you have 20 minutes to spare,
you can [watch a video of Zhian's UseR!2021 talk: "Using R as Community
Workbench for The Carpentries Lesson Infrastructure][yt-vid] (with [slides
available here](https://zkamvar.github.io/user2021/#1)).

> Because we deeply understand the implications of changing things too rapidly,

Reviewer: ```suggestion
Because these are large changes with far-reaching consequences for our community,
```

we are going to make this transition as smooth as possible, starting with **a

> soft beta release April 2022** with interested maintainers, lesson authors, and

Reviewer 2: ```suggestion
soft beta release in April 2022** with interested maintainers, lesson authors, and
```

> instructors. **Are you interested in helping out with testing the new
> infrastructure in how you work (either teaching, authoring, or maintianing)**

Reviewer: ```suggestion
instructors. **If you would like to begin exploring The Carpentries Workbench
and help us out with beta testing, please get in touch CONTACT INFO HERE.**
It is vital that we include a wide range of perspectives in this process.
We would like to learn from your experience with the new infrastructure 
however you work with our lesson sites: 
whether that is teaching the lesson in a workshop, 
developing a new lesson, 
or maintaining an existing repository.
Similarly, we are keen to include you regardless of your familiarity with the infrastructure on which our
lessons are currently based: 
regardless of whether you are new to The Carpentries community or 
have been working with our lesson sites for many years,
we want you to get involved.
```

> 

Reviewer: This is a long post, from which I would remove nothing, so I suggest to include a TOC here, with links for the reader to jump straight to the section they are most interested in.

Author: Done!

## Show me!

You can see The Carpentries Workbench in action at [the documentation 
site][workbench]. Of course, pictures are worth a thousand words. I have
converted the "Programming With Python" Software Carpentry Lesson and hosted it
in a sandbox github organisation[^note that this conversion was automated, so

> there may be a few rendering errors]. This is what the 'Making Choices' episode

Reviewer: ```suggestion
there may be a few rendering errors - please open an issue on the repository if you find any]. This is what the 'Making Choices' episode
```

  <figcaption>
Screencapture of <a href="https://swcarpentry.github.io/python-novice-inflammation/07-cond/">Programming With Python built with the Styles template</a> on 2022-01-27
  </figcaption>
</figure>

In this example repository, I show what it will look like when we implement [The
Carpentries Workbench][workbench]:

<figure style="text-align: center">
  <img src="{{site.urlimg }}/blog/2022/01/2022-01-31-python-choice.png" 
   alt="Screen capture of the 'Making Choices' episode in the Programming With Python lesson using The Carpentries Workbench. It has a toolbar at the top followed by a progress bar that indicates this episode is 55% through the lesson, and a sidebar with navigation that says CHAPTERS. The main content shows the episode title and a grey and blue overview box showing questions and objectives "/>
  <figcaption>
Screencapture of <a href="https://fishtree-attempt.github.io/python-novice-inflammation/07-cond">Programming With Python built with the Carpentries Workbench</a> on 2022-01-27
  </figcaption>

> </figure>

Reviewer: Can you use a screencapture that includes the equivalent content to that shown in the previous figure? E.g. I would like to be able to see the code blocks that are included in the _styles_ screengrab.

Author: done

Callout block from <a href="https://swcarpentry.github.io/python-novice-inflammation/01-intro/">Programming With Python built with the Styles template</a> on 2022-01-27
  </figcaption>
</figure>

<figure style="text-align: center">
  <img src="{{site.urlimg }}/blog/2022/01/2022-01-31-python-callout.png" 
   alt="Screen Capture of the first callout block from the introductory episode in the Programming with Python lesson using The Carpentries Workbench. It shows a grey box with a  dark grey border on the left and bell icon in a dark grey box and a title that says 'Variables as Sticky Notes'. The text is larger and there are three images and two codeblocks inside. The code blocks have thick lines running across the top that clearly separate them from the content."/>
  <figcaption>
Callout block from <a href="https://fishtree-attempt.github.io/python-novice-inflammation/01-intro">Programming With Python built with the Carpentries Workbench</a> on 2022-01-27
  </figcaption>
</figure>

For a full list of features in the episodes, you can visit the [Episode 
Structure](https://carpentries.github.io/sandpaper-docs/episodes.html) episode 
in the Carpentries Workbench documentation. 

> ### For Instructors: Instructor Notes

Reviewer: ```suggestion
### For Instructors: Instructor View
```

One resource that is useful for instructors teaching new-to-them lessons is the

> instructor notes, but this page is often buried in the "extras" dropdown menu

Reviewer 2: ```suggestion
instructor notes, but this page is often buried in the "extras" dropdown menu.
Because it is out of sight, it is not used often, and goes out of sync with the
content of the lesson.
```

> and while it is useful to have the overview, the new infrastructure offers a way

Reviewer 2: ```suggestion
And while it is useful to have an overview of all the tips to teach a lesson, the new infrastructure offers a way
```

to include these _in context_ in the lesson with the new _instructor view_.

First, visit <https://carpentries.github.io/sandpaper-docs/episodes#instructor-notes> and you will see a section that contains a heading, paragraph, and a discussion block (in yellow).

Now, visit <https://carpentries.github.io/sandpaper-docs/instructor/episodes#instructor-notes> and you will find an instructor note revealed:

<figure style="text-align: center">
  <img src="{{site.urlimg }}/blog/2022/01/2022-01-31-instructor-note.png" 
   alt="Screen capture of a section showing an embedded instructor note which appears as a light grey rectangle spanning the page with a magenta box containing a white pencil icon on the left and a downward-pointing chevron on the right indicating that it contains content."/>
  <figcaption>
Screencapture of an instructor note from <a href="https://carpentries.github.io/sandpaper-docs/episodes#instructor-notes">The Carpentries Workbench Documentation</a> on 2022-01-27
  </figcaption>
</figure>

Once we have the infrastructure established, we hope to organise instructor

> note sprints for our instructors to share their knowledge and contribute to the

Reviewer 2: ```suggestion
note sprints for our Instructors to share their knowledge and contribute to the
```

a pull request from the [styles repository][styles], but this can lead to a 
backlog of github notifications:

<figure style="text-align: center">
  <img src="{{ site.urlimg }}/blog/2022/01/2022-01-31-gh-notifications.png" 
   alt="List of notifications requesting an update of lessons with the styles template"/>
  <figcaption>
A huge backlog of notifications that Zhian needs to address :'(
  </figcaption>
</figure>

With the new lesson infrastructure, changes, even major changes are applied
immediately. Take for example this [chapter on editing (captured on 
2022-01-26)][new-episodes] and compare it with [the same content, but using the 
previous version of the lesson infrastructure][old-episodes]. This change
happened with the click of a button (or had I been patient, I could have just
waited a week for the change to implement itself). 

### And More!!!

There are many more features that should make authoring, contributing, and
maintaining lessons easier for everyone including automated checks for 
missing alt text, abnormal heading structure, and invalid internal links.

We are now piloting the inclusion of BioSchemas metadata in our

> lessons so that they can be indexed at [ELIXER's Training Portal, 
> TeSS](https://tess.elixir-europe.org/), which will give more exposure to the

Reviewer: ```suggestion
lessons so that they can be indexed and aggregated on repositories such as [ELIXIR's Training Portal, 
TeSS](https://tess.elixir-europe.org/), which will increase the lessons' reach and give more exposure to the
```

work of our lesson authors and maintainers.

> One of the things I'm really excited about with the lesson infrastructure is a

Reviewer: ```suggestion
One of the things we're really excited about with the lesson infrastructure is a
```

you were using "we" for most of the post up to now - recommend to stick with it unless it's you are really only talking about yourself (like with the aside about impatience, above)

[less invasive and faster build process for our R-based
lessons][sandpaper-cache] that freezes the package versions and provides [pull

> request previews][pr-preview] that will avoid lessons suddenly breaking when
> new package versions being released.
> 

Reviewer: ```suggestion
request previews][pr-preview] that will avoid lessons suddenly breaking when
new package versions are released.

```

### Demonstration

Zhian is going to implement and fill this in on Friday :weary:

## Transition Plan

> As I mentioned earlier, we will not make this transition all at once. There are

Reviewer: ```suggestion
As mentioned earlier, we will not make this transition all at once. There are
```

over a hundred lessons across The Carpentries (including Lab and Incubator),
each with their own unique setup. Before any transition will happen, we will do

> one final lesson release and then slowly migrate the lessons based on an opt-in

Reviewer: ```suggestion
one more lesson release and then slowly migrate the lessons based on an opt-in
```

> process by the maintainers and the instructors will be notified at least two

Reviewer: ```suggestion
process by the maintainers. Instructors will be notified at least two
```

> weeks in advance of the migration. 

Reviewer: ```suggestion
weeks in advance of these migrations. 
```

> For the maintainers, I have set up [a workflow for automated conversion with
> customisation](https://github.com/data-lessons/lesson-transition/), and once
> the maintainers/authors of the lesson are ready and they consent, I will
> transition the lesson, preserving the commit history (stripping the styles-
> specific commits). 

Reviewer: ```suggestion
For the maintainers, we have set up [a workflow for automated conversion with
customisation](https://github.com/data-lessons/lesson-transition/), and once
the maintainers/authors of the lesson are ready and they consent, we will
transition the lesson, preserving the commit history (stripping the styles-
specific commits). 
```

I feel bad about this one, because it really has been all your work and maybe using "we" dilutes the respect you have earned. Feel free to reject this suggestion.

Author: I don't feel bad about this at all! Less inherent responsibility for me!

## Future Plans

The Carpentries Workbench is the first step towards making our lesson
infrastructure more equitable and accessible, but we are far from finished. In
the future, we can envision a lesson infrastructure that drives lesson

> developers towards the pit of success in terms of creating accessible and

Reviewer: "pit of success" was a new phrase for me, and may not be easy from an i18n perspective - could you link to a resource where someone could understand better what you mean?

Author: That is a good point! I meant to link it. https://engineering.rice.edu/news/hadley-wickham-identifies-good-design-data-science

> developers towards the pit of success in terms of creating accessible and

Author: ```suggestion
developers towards the [pit of success](https://engineering.rice.edu/news/hadley-wickham-identifies-good-design-data-science) in terms of creating accessible and
```

> up-to-date lessons in any lanuguge (both human and computer).

Reviewer: ```suggestion
up-to-date lessons in any language (both human and computer).
```

Author: sruly I maent lanuguge

> up-to-date lessons in any lanuguge (both human and computer).

Reviewer 2: this paragraph about future plans might be a  good place to add a mention that the LIC will start their activities again as we get closer to release, and pointing to your blog post on the topic.

> OMAR PLEASE HELP THING OF A GOOD ENDING

Reviewer: You should definitely repeat the call to action at the end, in addition to a sentence or two to wrap everything up.

Reviewer 2: Also, we need a mention that this work is funded by CZI Science.

Author: Hai @kmomar 👋🏼

Reviewer 3: Add something like:

### How you can help 
- If you are interested in beta testing The Carpentries Workbench and providing valuable feedback as we iterate on the infrastructure, please contact zkamvar@carpentries.org
- If you find any errors in the infrastructure, please consider opening an issue on the repository (link) 
- Please continue providing feedback on the work outlined above 

Thank you to our community who has been instrumental in calling for these changes, and in advance for helping The Carpentries as we transition to a more equitable and accessible lesson infrastructure. Thank you also to [CZI Science](https://chanzuckerberg.com/science/) for their support and funding for this work.

Reviewer 4: Would you be interested in scheduling a Community Discussion for this? Seems it would be a good themed topic that we could get scheduled whenever is convenient for you. No worries if you prefer not to....just a thought.I can work with you to get it added to the Community Calendar and Etherpad and help pull something together so we get good feedback during the call

Author: Yes I would! We discussed these in the last Curriculum Team Meeting. I'm not sure when they should be, but we do know that they should be targetted to specific members of the community so that we can highlight and discuss aspects of the workbench that are relative for that community (e.g. while members of the Instructor community will be tangentially interested in the build process for the lessons, the main topics of focus will be the navigational features of the lessons, the instructor view, and future features such as handouts and/or slides).

[R](https://r-project.org) and [Pandoc](https://pandoc.org) in three custom
packages called [{sandpaper}], [{varnish}], and [{pegboard}] for deployment,
styling, and validation, respectively. It is important to note here that
knowledge of R or pandoc is not necessary to build a lesson using The
Workbench. The only requirement is a working knowledge of markdown syntax.
To find out more about the process we used for validating the usability of the
infrastructure, you can read our [blog post about the alpha testing][alpha-test]
and our [blog post about the UX testing and design][ux-test], or if you have 20 
minutes to spare, you can [watch a video of Zhian's UseR!2021 talk: "Using R as Community Workbench for The Carpentries Lesson Infrastructure][yt-vid]
(with [slides available here](https://zkamvar.github.io/user2021/#1)).

Because these are large changes with far-reaching consequences for our community,
we are going to make this transition as smooth as possible, starting with **a
soft beta release in April 2022** with interested maintainers, lesson authors, and

> instructors. **If you would like to begin exploring The Carpentries Workbench

Reviewer 4: Our style guide has us capitalise all volunteer roles so should check on that throughout

Author: Good catch! I'll fix that!

to grow.

Following the guidelines we set out in [our post on design principles][design-principles], The Carpentries Workbench is built with
[R](https://r-project.org) and [Pandoc](https://pandoc.org) in three custom
packages called [{sandpaper}], [{varnish}], and [{pegboard}] for deployment,
styling, and validation, respectively. It is important to note here that
knowledge of R or pandoc is not necessary to build a lesson using The
Workbench. The only requirement is a working knowledge of markdown syntax.
To find out more about the process we used for validating the usability of the
infrastructure, you can read our [blog post about the alpha testing][alpha-test]
and our [blog post about the UX testing and design][ux-test], or if you have 20 
minutes to spare, you can [watch a video of Zhian's UseR!2021 talk: "Using R as Community Workbench for The Carpentries Lesson Infrastructure][yt-vid]
(with [slides available here](https://zkamvar.github.io/user2021/#1)).

Because these are large changes with far-reaching consequences for our community,
we are going to make this transition as smooth as possible, starting with **a

> soft beta release in April 2022** with interested Maintainers, Lgsson Authors, and

Reviewer 2: ```suggestion
soft beta release in April 2022** with interested Maintainers, Lesson Authors, and
```

  </figcaption>
</figure>

Once we have the infrastructure established, we hope to organise Instructor
note sprints for our Instructors to share their knowledge and contribute to the
lesson development and maintenance.

### For Maintainers: Getting the Latest Features!

Our [current lesson template][styles] can be thought of as an all-in-one box of
tools that you can add markdown files to and get a lesson out of at the end. One
of the biggest pain points of this setup is that [as the template improves, older lessons rapidly become outdated and miss out on new features][slide-19].
The official way to update a lesson with the new features is to manually create
a pull request from the [styles repository][styles], but this can lead to a 
backlog of github notifications:

<figure style="text-align: center">
  <img src="{{ site.urlimg }}/blog/2022/01/2022-01-31-gh-notifications.png" 
   alt="List of notifications requesting an update of lessons with the styles template"/>
  <figcaption>
A huge backlog of notifications that Zhian needs to address :'(
  </figcaption>
</figure>

> With the new lesson infrastructure, changes, even major changes are applied

Reviewer 2: ```suggestion
With the new lesson infrastructure, changes to the styling, even major changes are applied
```

and compare it with [the same content, but using the previous version of the lesson infrastructure][old-episodes]. This change
happened with the click of a button (or had I been patient, I could have just
waited a week for the change to implement itself). 

### And More!!!

There are many more features that should make authoring, contributing, and
maintaining lessons easier for everyone including automated checks for 
missing alt text, abnormal heading structure, and invalid internal links.

We are now piloting the inclusion of BioSchemas metadata in our
lessons so that they can be indexed and aggregated on repositories such as 
[ELIXIR's Training Portal, TeSS](https://tess.elixir-europe.org/), which will
increase the lessons' reach and give more exposure to the
work of our Lesson Authors and Maintainers.

One of the things we're really excited about with the lesson infrastructure is a
[less invasive and faster build process for our R-based lessons][sandpaper-cache]
that freezes the package versions and provides [pull request previews][pr-preview] 
that will avoid lessons suddenly breaking when new package versions are released.

### Demonstration

> Zhian is going to implement and fill this in on Friday :weary:

Reviewer 2: update or remove section?

Author: let's remove it

 - Please continue providing feedback on the work outlined above

Thank you to our community who has been instrumental in calling for these
changes, and in advance for helping The Carpentries as we transition to a more
equitable and accessible lesson infrastructure. Thank you also to 
[CZI Science](https://chanzuckerberg.com/science/)
for their support and funding for this work.

[design-principles]: https://carpentries.org/blog/2020/08/lesson-template-design/
[{sandpaper}]: https://carpentries.github.io/sandpaper/
[{pegboard}]: https://carpentries.github.io/pegboard/
[{varnish}]: https://github.com/carpentries/varnish#readme
[alpha-test]: https://carpentries.org/blog/2021/07/infrastructure-testing/
[ux-test]: https://carpentries.org/blog/2021/05/lesson-template-design-process/
[yt-vid]: https://www.youtube.com/watch?v=vd8XZSuY_Rs&list=PLSFzyC3wp8-csb8rtreOUoW8C_1J87QD5&index=1&t=1271s
[workbench]: https://carpentries.github.io/sandpaper-docs
[styles]: https://github.com/carpentries/styles/
[slide-19]: https://zkamvar.github.io/user2021/#19
[old-episodes]: https://web.archive.org/web/20220125163344/https://carpentries.github.io/sandpaper-docs/episodes.html

> [new-episodes]: https://web.archive.org/web/20220127011238/https://carpentries.github.io/sandpaper-docs/episodes.html

Reviewer 2: this link does not work (it says "The Wayback Machine has not archived that URL.")

Author: ```suggestion
[new-episodes]: https://carpentries.github.io/sandpaper-docs/episodes.html
```
